### ISSUE TITLE
Deploy p2 repository artifacts to maven central.

### ISSUE THREAD

**Issue by nikhilnanivadekar** [eclipse-collections/eclipse-collections#294]

**Comment by guw** [comment]
`mvn deploy` inside `p2-repository/org.eclipse.collections` should deploy it to the configured Maven repository.

**Comment by nikhilnanivadekar** [comment]
@guw the deploy step inside p2-repository/org.eclipse.collections fails due to jar signer. So I guess this and #287 are related. One fixes the other.

```
[ERROR] Failed to execute goal org.eclipse.cbi.maven.plugins:eclipse-jarsigner-plugin:1.1.3:sign (jarsign) on project org.eclipse.collections: Execution jarsign of goal org.eclipse.cbi.maven.plugins:eclipse-jarsigner-plugin:1.1.3:sign failed: Invalid signature file digest for Manifest main attributes -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.cbi.maven.plugins:eclipse-jarsigner-plugin:1.1.3:sign (jarsign) on project org.eclipse.collections: Execution jarsign of goal org.eclipse.cbi.maven.plugins:eclipse-jarsigner-plugin:1.1.3:sign failed: Invalid signature file digest for Manifest main attributes
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution jarsign of goal org.eclipse.cbi.maven.plugins:eclipse-jarsigner-plugin:1.1.3:sign failed: Invalid signature file digest for Manifest main attributes
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.lang.SecurityException: Invalid signature file digest for Manifest main attributes
	at sun.security.util.SignatureFileVerifier.processImpl(SignatureFileVerifier.java:314)
	at sun.security.util.SignatureFileVerifier.process(SignatureFileVerifier.java:268)
	at java.util.jar.JarVerifier.processEntry(JarVerifier.java:316)
	at java.util.jar.JarVerifier.update(JarVerifier.java:228)
	at java.util.jar.JarInputStream.read(JarInputStream.java:212)
	at java.util.zip.ZipInputStream.closeEntry(ZipInputStream.java:140)
	at java.util.zip.ZipInputStream.getNextEntry(ZipInputStream.java:118)
	at java.util.jar.JarInputStream.getNextEntry(JarInputStream.java:142)
	at java.util.jar.JarInputStream.getNextJarEntry(JarInputStream.java:179)
	at org.eclipse.cbi.maven.plugins.jarsigner.JarResigner.isAlreadySigned(JarResigner.java:118)
	at org.eclipse.cbi.maven.plugins.jarsigner.JarResigner.sign(JarResigner.java:94)
	at org.eclipse.cbi.maven.plugins.jarsigner.RecursiveJarSigner.doSignJarRecursively(RecursiveJarSigner.java:64)
	at org.eclipse.cbi.maven.plugins.jarsigner.RecursiveJarSigner.doSignJar(RecursiveJarSigner.java:41)
	at org.eclipse.cbi.maven.plugins.jarsigner.FilteredJarSigner.sign(FilteredJarSigner.java:44)
	at org.eclipse.cbi.maven.plugins.jarsigner.mojo.SignMojo.signArtifact(SignMojo.java:298)
	at org.eclipse.cbi.maven.plugins.jarsigner.mojo.SignMojo.execute(SignMojo.java:280)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
```

**Comment by guw** [comment]
There are some subtle issues around jar signing - especially timing, i.e. p2 modifies a JAR adding metadata which could invalidate its signature. It gets worse when pack200 is done, too.

There is a good blog here:
http://www.codetrails.com/blog/sign-your-eclipse-project

**Comment by nikhilnanivadekar** [comment]
So, should we remove the jar signer altogether and use pack200 instead everywhere? As it is we are on eclipse's hudson instance. Also, do I need to add the signer explicitly in p2 module?

**Comment by guw** [comment]
So the p2 module should not inherit signing from the global parent. It should define its own signing+pack200 steps as show in the blog.

**Comment by nikhilnanivadekar** [comment]
That might be tricky since the jar signer seems to sign every artifact built as a part of deploy step. I'll try out tomorrow. If you get some spare time, I will take a PR for this :)

**Comment by nikhilnanivadekar** [comment]
@guw I need help with the jar signer. I have tried everything which the blog said, as well followed the documentation about jar signer. However, the deploy fails. Is there a way I can test this in my local machine? I am unable to understand what I am doing wrong.

https://hudson.eclipse.org/collections/job/deploy/76/console

**Comment by guw** [comment]
That message looks odd. Any chance that something in the build process already signed the jar?

**Comment by nikhilnanivadekar** [comment]
@guw The parent jar-signer runs during `verify` phase: https://github.com/eclipse/eclipse-collections/blob/master/pom.xml#L661

Where as the p2 repo's jar-signer runs during `package` phase: https://github.com/eclipse/eclipse-collections/blob/master/p2-repository/pom.xml#L140

Both the jar-signers are triggered only when build is running with `release-artifacts` profile. From my understanding (which can be incorrect), the other modules will be signed during `verify` phase while p2 will be signed during `package` phase.
Which is the correct phase for the jar-signer? Can we make everything consistent? 
Also, is there a way to locally test these changes?

**Comment by guw** [comment]
The parent jar-signer must not run at all. Can this be disabled?

**Comment by nikhilnanivadekar** [comment]
@guw do you mean not run for p2 or not run for the library in total?
We didn't sign our jars before, but we were advised to start signing them. 
I think the easiest way forward is to have a new build which will deploy p2 artifacts to sonatype. That jar-signer will be used only for p2. So it gives us flexibility to sign whenever we want and whatever we want.
Thoughts?

**Comment by guw** [comment]
It must not run for the p2 build. It can run for the regular build. The p2 build signs the jars itself using special processing to allow embedding additional metadata.

**Comment by nikhilnanivadekar** [comment]
Ok makes sense. It would be better to have another Hudson job then to push to p2. Will try it on Friday for Milestone 3.
Thanks a lot for your help!

**Comment by nikhilnanivadekar** [comment]
@guw I am hopeful that this will work. Please let me know if you see something wrong.

1) For Maven Central:
https://github.com/eclipse/eclipse-collections/pull/327/commits/c1c769285b699e352bbb13265492469cc13e0c71#diff-bb39bc4449e09dadf2be960bdd8dcac0R113
Along with a new build with config:
```
-pl
'!jmh-scala-tests,!jmh-tests,!junit-trait-runner,!scala-unit-tests,!serialization-tests,!test-coverage,!unit-tests-java8,!unit-tests,!p2-repository,!p2-repository/org.eclipse.collections'
-DperformReleaseP2=true
-Dsurefire.useFile=false
clean
javadoc:jar
install
eclipse-jarsigner:sign
deploy
```

2) For Eclipse Marketplace:
https://github.com/eclipse/eclipse-collections/pull/327/commits/c1c769285b699e352bbb13265492469cc13e0c71#diff-bb39bc4449e09dadf2be960bdd8dcac0R212
```
-pl
'!jmh-scala-tests,!jmh-tests,!junit-trait-runner,!scala-unit-tests,!serialization-tests,!test-coverage,!unit-tests-java8,!unit-tests'
-DpublishP2=true
-Dsurefire.useFile=false
clean
install
eclipse-jarsigner:sign
```

**Comment by motlin** [comment]
@itohro we've always signed our jars, haven't we?

**Comment by nikhilnanivadekar** [comment]
@guw The p2 jar signer still fails now this time with a new exception, so I guess we are making some progress.
Hudson build: https://hudson.eclipse.org/collections/job/publish-p2-repo/4/console

Also raised a Bugzilla: https://bugs.eclipse.org/bugs/show_bug.cgi?id=521389

```
[ERROR] Failed to execute goal org.eclipse.tycho.extras:tycho-pack200a-plugin:1.0.0:normalize (pack200-normalize) on project org.eclipse.collections: Could not pack200 normalize jar file /jobs/genie.collections/publish-p2-repo/workspace/p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0.M3.jar: pack200:normalize cannot be called for signed jar /jobs/genie.collections/publish-p2-repo/workspace/p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0.M3.jar -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho.extras:tycho-pack200a-plugin:1.0.0:normalize (pack200-normalize) on project org.eclipse.collections: Could not pack200 normalize jar file /jobs/genie.collections/publish-p2-repo/workspace/p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0.M3.jar
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Could not pack200 normalize jar file /jobs/genie.collections/publish-p2-repo/workspace/p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0.M3.jar
	at org.eclipse.tycho.extras.pack200.mojo.Pack200NormalizeMojo.execute(Pack200NormalizeMojo.java:103)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.io.IOException: pack200:normalize cannot be called for signed jar /jobs/genie.collections/publish-p2-repo/workspace/p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0.M3.jar
	at org.eclipse.tycho.extras.pack200.Pack200Archiver.normalize(Pack200Archiver.java:55)
	at org.eclipse.tycho.extras.pack200.mojo.Pack200NormalizeMojo.execute(Pack200NormalizeMojo.java:89)
	... 22 more
```

**Comment by nikhilnanivadekar** [comment]
@motlin the question here is not about signing the jars. Yes we sign the jars for our Maven Central (m2 repository) artifacts. However, for OSGi bundles the sign process is different for the p2 repository. I might be wrong, but I am still learning.

**Comment by leomrlima** [comment]
@nikhilnanivadekar, 

9.0.0.M3 error:
`An error occurred while collecting items to be installed

session context was:(profile=epp.package.jee, phase=org.eclipse.equinox.internal.p2.engine.phases.Collect, operand=, action=).

Artifact not found: http://download.eclipse.org/collections/9.0.0.M3/repository/plugins/org.eclipse.collections_9.0.0.v20170824-1703.jar.

http://download.eclipse.org/collections/9.0.0.M3/repository/plugins/org.eclipse.collections_9.0.0.v20170824-1703.jar

Artifact not found: http://download.eclipse.org/collections/9.0.0.M3/repository/plugins/org.eclipse.collections.source_9.0.0.v20170824-1703.jar.

http://download.eclipse.org/collections/9.0.0.M3/repository/plugins/org.eclipse.collections.source_9.0.0.v20170824-1703.jar`

8.2.0 unsigned:
![screen shot 2017-08-24 at 1 14 03 pm](https://user-images.githubusercontent.com/1250061/29681756-52c8ca84-88ce-11e7-812d-6c8db1c4d095.png)

**Comment by nikhilnanivadekar** [comment]
@guw from @leomrlima 's message, seems like when the pack2000 and tycho plugins are triggered we are missing some artifacts. At least the jar is getting signed in the p2 build now. Any insights on what is missing, and why it could be missing?

**Comment by leomrlima** [comment]
@nikhilnanivadekar , EC9M3 not signed:
![screen shot 2017-08-24 at 2 06 43 pm](https://user-images.githubusercontent.com/1250061/29683870-841f872e-88d5-11e7-8924-0a19130dd556.png)

**Comment by nikhilnanivadekar** [comment]
Thanks @leomrlima !
The problem is down to the jar signer inside the p2.

**Comment by guw** [comment]
Sorry for the long delay. As commented on the bug, the pack200:normalize task needs to be run on a non-signed jar.

https://wiki.eclipse.org/Tycho/Pack200#Pack200_and_Signing

1. pack200a:normalize
2. sign
3. pack200b:pack

**Comment by nikhilnanivadekar** [comment]
@guw I thought I already incorporated that: https://github.com/eclipse/eclipse-collections/blob/master/p2-repository/pom.xml#L209
Can you please take a look?

Is that not correct? Can you please guide me to a sample project where this is done and works fine?

**Comment by guw** [comment]
@nikhilnanivadekar It looks correct. However, the error message clearly says that the jar is already signed at the normalize stage. Can you verify that the regular jar signer (from your Maven parent) is not running at all? Also, the `resigningStrategy` configuration shouldn't be necessary. What happens when you remove it?

**Comment by ujhelyiz** [comment]
I have looked into the building code, and I think I have found out what the issue is: in short, instead of relying the custom `eclipse-sign-p2` profile, simply set up the property `-DsigningServiceType=eclipse` for signing to work as expected.

----
A more detailed analysis:

The `eclipse-sign-p2` profile includes direct setup of the build process does not sign and repack the jars included in the update site but only the one in `p2-repository/org.eclipse.collections/target/org.eclipse.collections-9.0.0-SNAPSHOT.jar` - the reason for this is that the ebr configuration runs earlier and already creates the p2 repository; on the other hand, the signing configuration does not select the jar files in the build p2 repository.

Luckily, EBR already supports direct signing with pack200 as described [here](https://github.com/eclipse/ebr/#pack200-and-signing-bundles) using the `signingServiceType` property. The use of this resulted in some warnings of not having a dedicated Tycho setup, so it fall back to Tycho 0.26.0, otherwise this setting seemed to work as expected.

I have set up such a temporary build configuration to test that [in the VIATRA HIPP](https://hudson.eclipse.org/viatra/job/ec-build-temporary/); you can look at the exact configuration by logging in with your Eclipse credentials. The built repository is also included as a build artifact available for testing.

**Comment by nikhilnanivadekar** [comment]
@ujhelyiz I do not  see the build configuration even when I login using Eclipse credentials. Would you mind copy pasting the config here or giving a screenshot? I will test it out on our snapshot builds.

**Comment by ujhelyiz** [comment]
Sorry for the configuration issue; it is strange because I could see your configuration at https://hudson.eclipse.org/collections/job/publish-p2-repo (in a read-only fashion, but it was helpful setting up an almost similar setting). I have now also tried to add you email address, so now it definitely should work now.

Regardless of that, I have also created a screenshot of the relevant part of the configuration:
<img width="950" alt="hudson-config" src="https://user-images.githubusercontent.com/251892/30395413-59410f5e-98c6-11e7-9894-0b5ff2ca67e3.png">

Basically, most changes are related to formatting (I moved the -D elements to the dedicated properties field to make it more readable). What is important to note, that (a) I have not enabled any specific profiles manually, and (b) I have defined the `signingService=eclipse` property.

**Comment by nikhilnanivadekar** [comment]
Yup I could see the config and I ran the build: https://hudson.eclipse.org/collections/job/publish-p2-repo/9/

Would you mind testing it out? 
http://download.eclipse.org/collections/9.0.0-SNAPSHOT/repository

**Comment by ujhelyiz** [comment]
The repository seems to work for to me, and it did not warn me for unsigned content.

**Comment by nikhilnanivadekar** [comment]
Perfect! Thank you very much. Should we then delete the profile? It will confuse others if it is there.
If we should delete the profile, would you like to make the contribution?
Thanks a lot for your help!

**Comment by ujhelyiz** [comment]
I have opened PR #350 that removes the profile.

**Comment by nikhilnanivadekar** [comment]
Thank you very much. Just want @guw to review if possible since I am not an OSGi expert. Would like at least 1 person to review it before I merge.

**Comment by nikhilnanivadekar** [comment]
Starting 9.1.0 should we start pushing to maven central? What do you think @ujhelyiz , @guw ?
I tried this during 9.0.0 but was unsuccessful. Would either of you have any insights on what should be done? The failure was consistently either during the SonaType upload or during jar sign. So, p2 repo build was skipped during the deploy build.

**Comment by guw** [comment]
I'm not sure about the benefit. Why do you want to deploy the p2 repo on Maven Central when it's available as a p2 repo from download.eclipse.org already? The p2 repo can be deployed as a whole zip.

**Comment by ujhelyiz** [comment]
I second @guw - I see little benefit in publishing the p2 repository to maven central. A zipped and an online version (unzipped into a folder) of the repository available from download.eclipse.org should be more than enough for every use case I have ever witnessed.

**Comment by nikhilnanivadekar** [comment]
How and where do I push the zipped repo? I think we are already pushing the unzipped repo.
Also, how can I make the repo discoverable? Seems like users might have trouble finding the repo. Or what we are doing now is pretty much standard and is discoverable?

**Comment by ujhelyiz** [comment]
For Eclipse plug-ins it is a widely accepted way for installation to look for the URL of the unzipped p2 repository in the project website. As of now, is that it is quite hard to find the repository URL from your website (http://www.eclipse.org/collections); the repository URL is only available in the release notes. But adding the URL should solve this issue. Providing a zipped repository use case is relevant where offline installation is a requirement for one of your users.

In such case, you should only push a zipped version of your generated p2 repository to a specific folder in download.eclipse.org (similar to the unzipped version). E.g. in case of the VIATRA project, the latest (1.6.1) release is available from http://download.eclipse.org/viatra/updates/release/1.6.1, and we also have a zipped site available at http://download.eclipse.org/viatra/updates/zipped/org.eclipse.viatra.update.1.6.1.zip (technically, the link from our download page uses the mirror url from eclipse.org, but that points to the given url).

**Comment by nikhilnanivadekar** [comment]
closing this as wont fix.
